### ISSUE TITLE
Two versions of GHC depended on when useArchiveFilesForTemplateHaskell is true

### ISSUE THREAD

**Issue by jonathanlking** [nh2/static-haskell-nix#122]
On master (88f1e2d) when setting `useArchiveFilesForTemplateHaskell = true` two versions of GHC are being depended on. Picking an arbitrary package (in this case `vector`):

```
nix-repl> pkgs = (import ./survey/default.nix { useArchiveFilesForTemplateHaskell = true; })

nix-repl> pkgs.haskellPackages.vector                                                        
«derivation /nix/store/v6hbyn8a5nydw81a9vjabr4qxfj093mr-vector-0.12.3.1.drv»

nix-repl> pkgs = (import ./survey/default.nix { })                                           

nix-repl> pkgs.haskellPackages.vector              
«derivation /nix/store/ggxx6mgws1vn22l668dsnpl76yx635ri-vector-0.12.3.1.drv»
```

Using `nix-tree` to inspect the derivations we find that the first derivation `/nix/store/v6hbyn8a5nydw81a9vjabr4qxfj093mr-vector-0.12.3.1.drv` depends on two different GHC derivations:

a. `/nix/store/6yvclb4xwmkwg632nmm2mdcaw94vfj7k-ghc-musl-9.2.7.drv` — Immediate Parents (41)
b. `/nix/store/pd3h7k2f9p9290dhfch2wnwkp9714xgk-ghc-musl-9.2.7.drv` — Immediate Parents (4): jailbreak-cabal-1.4.drv, hscolour-1.24.4.drv, Cabal-syntax-3.6.0.0.drv, hscolour-1.24.4.drv

While the second derivation `/nix/store/ggxx6mgws1vn22l668dsnpl76yx635ri-vector-0.12.3.1.drv` _only_ depends on a.

Diffing the derivations for a and b, the only meaningful difference is in the `preConfigure` step:

```diff
- DYNAMIC_GHC_PROGRAMS = YES
+ DYNAMIC_GHC_PROGRAMS = NO
+ GhcLibHcOpts += -fPIC -fexternal-dynamic-refs
+ GhcRtsHcOpts += -fPIC -fexternal-dynamic-refs
```

Some observations/thoughts:

1. These are options changed by `enableRelocatedStaticLibs` and `enableShared` attributes in the [GHC derivation](https://github.com/NixOS/nixpkgs/blob/nixos-23.05/pkgs/development/compilers/ghc/9.4.4.nix), which is set by `useArchiveFilesForTemplateHaskell`.
2. The immediate parents suggest this is the GHC in `buildHaskellPackages`, as these parents match things set up in [`mkDerivationImpl`](https://github.com/NixOS/nixpkgs/blob/nixos-23.05/pkgs/development/haskell-modules/make-package-set.nix#L50).
3. We already override `buildHaskellPackages` to replace GHC (to avoid depending on 2 versions) https://github.com/nh2/static-haskell-nix/blob/88f1e2d57e3f4cd6d980eb3d8f99d5e60040ad54/survey/default.nix#L1586-L1596.
4. The GHC overrides are made by a [`fixGhc`](https://github.com/nh2/static-haskell-nix/blob/88f1e2d57e3f4cd6d980eb3d8f99d5e60040ad54/survey/default.nix#L1562) function, which is applied to both `ghc` and the `ghc` in `buildHaskellPackages`, so I'm not sure why it isn't working the same in both places?
5. `buildHaskellPackages.ghc` _does_ have the right derivation 🤔

    ```
    nix-repl> pkgs.haskellPackages.buildHaskellPackages.ghc
    «derivation /nix/store/6yvclb4xwmkwg632nmm2mdcaw94vfj7k-ghc-musl-9.2.7.drv»

    nix-repl> pkgs.haskellPackages.ghc                      
    «derivation /nix/store/6yvclb4xwmkwg632nmm2mdcaw94vfj7k-ghc-musl-9.2.7.drv»
    ```

**Comment by jonathanlking** [comment]
I intend to continue investigating this and will share thoughts here 📝
It's also very possible that I'm making a very silly mistake somewhere 😅

**Comment by jonathanlking** [comment]
Commenting out https://github.com/nh2/static-haskell-nix/blob/88f1e2d57e3f4cd6d980eb3d8f99d5e60040ad54/survey/default.nix#L1594-L1596 I now get 

```
nix-repl> pkgs = (import ./survey/default.nix { useArchiveFilesForTemplateHaskell = true; })

nix-repl> pkgs.haskellPackages.buildHaskellPackages.ghc                     
«derivation /nix/store/pd3h7k2f9p9290dhfch2wnwkp9714xgk-ghc-musl-9.2.7.drv»

nix-repl> pkgs.haskellPackages.ghc                      
«derivation /nix/store/6yvclb4xwmkwg632nmm2mdcaw94vfj7k-ghc-musl-9.2.7.drv»
```

So maybe this override isn't being applied "deep"/early enough? (So that it also applies to the packages within `haskellPackages` and not just to the exposed `buildHaskellPackages`?)

**Comment by jonathanlking** [comment]
Unfortunately I'm pretty stumped on this 😔

For posterity, one approach to debugging I tried was:
* Cloning a copy of nixpkgs (to `~/github/nixpkgs`)
* Adding a breakpoint to [`mkDerivation`](https://github.com/NixOS/nixpkgs/blob/312ab59e8ade69e6083017bd9b98a2919f1fb86a/pkgs/development/haskell-modules/make-package-set.nix#L74) with `mkDerivation = (_: break _) (makeOverridable mkDerivationImpl);`
* Launching `nix repl --debugger --ignore-try`
* Loading with my local modified nixpkgs `pkgs = (import ./survey/default.nix { normalPkgs = (import ~/github/nixpkgs/default.nix {}); useArchiveFilesForTemplateHaskell = true; })`
* Building the packages again with `pkgs.haskellPackages.vector`, which triggers the breakpoint (multiple times!)
* Printing out values of `ghc`/`self.ghc`/`buildHaskellPackages.ghc`/`self.buildHaskellPackages.ghc` that are in scope.

I didn't really learn anything new from this though, **so I intend to stop investigating for now**.
I _think_ the only consequence of this issue are longer build times/more disk space, due to two copies of GHC, rather than bad build output.

**Comment by jonathanlking** [comment]
Fixed by #127 (which incorporated #123).
